core/src/main/scala/kafka/coordinator/GroupCoordinator.scala

```diff
@@ -0,0 +1,511 @@
+                      memberId: String,
+                      sessionTimeoutMs: Int,
+                      protocolType: String,
+                      protocols: List[(GroupProtocol, Array[Byte])],
+                      responseCallback: JoinCallback) {
+    if (!isActive.get) {
+      responseCallback(Map.empty, memberId, 0, GroupCoordinator.NoProtocol, Errors.GROUP_COORDINATOR_NOT_AVAILABLE.code)
+    } else if (!isCoordinatorForGroup(groupId)) {
+      responseCallback(Map.empty, memberId, 0, GroupCoordinator.NoProtocol, Errors.NOT_COORDINATOR_FOR_GROUP.code)
+    } else if (sessionTimeoutMs < groupConfig.groupMinSessionTimeoutMs ||
+               sessionTimeoutMs > groupConfig.groupMaxSessionTimeoutMs) {
+      responseCallback(Map.empty, memberId, 0, GroupCoordinator.NoProtocol, Errors.INVALID_SESSION_TIMEOUT.code)
+    } else {
+      // only try to create the group if the group is not unknown AND
+      // the member id is UNKNOWN, if member is specified but group does not
+      // exist we should reject the request
+      var group = coordinatorMetadata.getGroup(groupId)
+      if (group == null) {
+        if (memberId != JoinGroupRequest.UNKNOWN_MEMBER_ID) {
+          responseCallback(Map.empty, memberId, 0, GroupCoordinator.NoProtocol, Errors.UNKNOWN_MEMBER_ID.code)
+        } else {
+          group = coordinatorMetadata.addGroup(groupId, protocolType)
+          doJoinGroup(group, memberId, sessionTimeoutMs, protocols, responseCallback)
+        }
+      } else if (group.protocolType != protocolType) {
+        responseCallback(Map.empty, memberId, 0, GroupCoordinator.NoProtocol, Errors.INCONSISTENT_GROUP_PROTOCOL.code)
+      } else {
+        doJoinGroup(group, memberId, sessionTimeoutMs, protocols, responseCallback)
+      }
+    }
+  }
+
+  private def doJoinGroup(group: GroupMetadata,
+                          memberId: String,
+                          sessionTimeoutMs: Int,
+                          protocols: List[(GroupProtocol, Array[Byte])],
+                          responseCallback: JoinCallback) {
+    group synchronized {
+      if (group.is(Dead)) {
+        // if the group is marked as dead, it means some other thread has just removed the group
+        // from the coordinator metadata; this is likely that the group has migrated to some other
+        // coordinator OR the group is in a transient unstable phase. Let the member retry
+        // joining without the specified member id,
+        responseCallback(Map.empty, memberId, 0, GroupCoordinator.NoProtocol, Errors.UNKNOWN_MEMBER_ID.code)
+      } else if (memberId != JoinGroupRequest.UNKNOWN_MEMBER_ID && !group.has(memberId)) {
+        // if the member trying to register with a un-recognized id, send the response to let
+        // it reset its member id and retry
+        responseCallback(Map.empty, memberId, 0, GroupCoordinator.NoProtocol, Errors.UNKNOWN_MEMBER_ID.code)
+      } else if (!group.supportsProtocols(protocols.map(_._1).toSet)) {
```

Comment:
That is right. If you attempt to upgrade the protocol without providing both versions, then the new group members (with the new protocol) would be rejected until all old members have left. This is consistent with the current implementation. I proposed previously to have the coordinator choose the protocol which the largest number of members support, but you seemed concerned that this would effectively halve the group's capacity for a short duration in a rolling upgrade. I still think that might be a more useful update mechanism in practice, but the implementation might be tricky since rejected members would have to retry at a later time which could lead to unnecessary (and costly) rebalancing. The advantage of the approach implemented here is that it is easy to implement.